File: app/src/main/java/fansirsqi/xposed/sesame/task/antForest/WhackMole.java
package fansirsqi.xposed.sesame.task.antForest;

import android.annotation.SuppressLint;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

import fansirsqi.xposed.sesame.util.GlobalThreadPools;
import fansirsqi.xposed.sesame.util.Log;
import fansirsqi.xposed.sesame.util.ResChecker;

/**
 * 6秒拼手速打地鼠（多线程+间隔控制版）
 * 同时开启6局游戏，结算能量最高的一局
 * 
 * @author Ghostxx
 * @date 2025/12/14
 */
public class WhackMole {
    private static final String TAG = WhackMole.class.getSimpleName();
    
    // 总游戏局数，同时开启3局游戏
    private static final int TOTAL_ROUNDS = 3;
    
    // 单个游戏任务的超时时间（秒），防止任务卡住
    private static final long TASK_TIMEOUT_SECONDS = 30;

    // 游戏会话信息
        private record GameSession(String token, List<String> remainingMoleIds, int whackedEnergy, int roundNumber) {
    }

    /**
     * 开6局游戏打地鼠（并发执行）
     * 每局游戏独立计时，严格控制在6秒内完成
     */
    @SuppressLint("DefaultLocale")
    public static void startWhackMole() {
        String source = "senlinguangchangdadishu";
        List<Future<GameSession>> futures = new ArrayList<>();
        
        // 创建自定义线程池
        ThreadPoolExecutor executor = new ThreadPoolExecutor(
            TOTAL_ROUNDS, TOTAL_ROUNDS,
            0L, TimeUnit.MILLISECONDS,
            new LinkedBlockingQueue<>(),
            new ThreadFactory() {
                private final AtomicInteger threadNumber = new AtomicInteger(1);
                @Override
                public Thread newThread(Runnable r) {
                    Thread t = new Thread(r, "WhackMole-" + threadNumber.getAndIncrement());
                    t.setDaemon(true);
                    return t;
                }
            }
        );
        
        try {
            // 1. 并发提交6局游戏任务
            Log.runtime(TAG, "启动" + TOTAL_ROUNDS + "局打地鼠游戏（并发模式）");
            for (int round = 1; round <= TOTAL_ROUNDS; round++) {
                final int currentRound = round;
                Callable<GameSession> task = () -> startWhackMole(currentRound, source);
                futures.add(executor.submit(task));
            }
            
            // 2. 等待所有任务完成
            List<GameSession> sessions = new ArrayList<>();
            for (int i = 0; i < futures.size(); i++) {
                Future<GameSession> future = futures.get(i);
                try {
                    GameSession session = future.get(TASK_TIMEOUT_SECONDS, TimeUnit.SECONDS);
                    if (session != null) {
                        sessions.add(session);
                        Log.runtime(TAG, String.format("第%d局完成，打地鼠能量 %dg", 
                                      session.roundNumber, session.whackedEnergy));
                    }
                } catch (TimeoutException e) {
                    Log.runtime(TAG, "第" + (i + 1) + "局任务超时");
                    future.cancel(true);
                } catch (Exception e) {
                    Log.runtime(TAG, "第" + (i + 1) + "局任务异常: " + e.getMessage());
                }
            }
            
            if (sessions.isEmpty()) {
                Log.runtime(TAG, "没有成功进行任何游戏");
                return;
            }
            
            // 3. 找出能量最高的局
            GameSession bestSession = sessions.get(0);
            for (GameSession session : sessions) {
                if (session.whackedEnergy > bestSession.whackedEnergy) {
                    bestSession = session;
                }
            }
            
            Log.runtime(TAG, String.format("最佳局为第%d局，打地鼠能量 %dg", 
                          bestSession.roundNumber, bestSession.whackedEnergy));
            
            // 4. 结算能量最高的那一局
            settleBestRound(bestSession, source);
            
        } catch (Exception e) {
            Log.runtime(TAG, "whackMole 任务调度失败: " + e.getMessage());
            Log.printStackTrace(TAG, e);
        } finally {
            shutdownExecutor(executor);
        }
    }
    
    /**
     * 进行单局游戏（严格控制在6秒内）
     */
    private static GameSession startWhackMole(int round, String source) {
        GlobalThreadPools.sleepCompat((long)(Math.random() * 250)); // 随机小延迟，错开请求时间
        long startTime = System.currentTimeMillis();
        try {
            Log.runtime(TAG, "第" + round + "局游戏开始");
            // 开始游戏
            JSONObject response = new JSONObject(AntForestRpcCall.startWhackMole(source));

            if (!ResChecker.checkRes(TAG + "拼手速游戏启动失败:", response)) {
                return null;
            }
            JSONArray moleInfoArray = response.optJSONArray("moleInfo");
            if (moleInfoArray == null || moleInfoArray.length() == 0) {
                Log.runtime(TAG, "第" + round + "局没有地鼠信息");
                return null;
            }

            String token = response.optString("token");
            // 收集地鼠信息
            List<Long> allMoleIds = new ArrayList<>();
            List<Long> bubbleMoleIds = new ArrayList<>();
            
            for (int i = 0; i < moleInfoArray.length(); i++) {
                JSONObject mole = moleInfoArray.getJSONObject(i);
                long moleId = mole.getLong("id");
                allMoleIds.add(moleId);
                if (mole.has("bubbleId")) {
                    bubbleMoleIds.add(moleId);
                }
            }
            
            Log.runtime(TAG, "第" + round + "局发现" + bubbleMoleIds.size() + "个能量地鼠");
            
            // 打地鼠
            int totalEnergy = 0;
            for (Long moleId : bubbleMoleIds) {
                try {
                    JSONObject whackResp = new JSONObject(AntForestRpcCall.whackMole(moleId, token, source));
                    if (whackResp.optBoolean("success")) {
                        int energy = whackResp.optInt("energyAmount", 0);
                        totalEnergy += energy;
                        Log.runtime(TAG, "第" + round + "局击打地鼠" + moleId + " 能量+" + energy + "g");
                    }
                } catch (Exception e) {
                    Log.runtime(TAG, "第" + round + "局打地鼠异常 " + moleId + ": " + e.getMessage());
                }
            }
            
            // 计算剩余未打的地鼠
            List<String> remainingMoleIds = new ArrayList<>();
            for (Long moleId : allMoleIds) {
                if (!bubbleMoleIds.contains(moleId)) {
                    remainingMoleIds.add(String.valueOf(moleId));
                }
            }
            
            // 等待接近6秒总时长（严格控制在6秒内）
            long elapsedTime = System.currentTimeMillis() - startTime;
            long sleepTime = Math.max(0, 6000 - elapsedTime - 300); // 提前300ms结束，确保不超标
            if (sleepTime > 0) {
                GlobalThreadPools.sleepCompat(sleepTime);
            }
            
            // 打印该局能量信息到森林日志
            Log.forest("森林能量⚡️[6秒拼手速第" + round + "局 打地鼠能量+" + totalEnergy + "g]");
            
            Log.forest(TAG, "第" + round + "局完成，总耗时" + (System.currentTimeMillis() - startTime) + "ms");
            
            // 返回会话信息
            return new GameSession(token, remainingMoleIds, totalEnergy, round);
            
        } catch (Exception e) {
            Log.runtime(TAG, "第" + round + "局游戏异常: " + e.getMessage());
            Log.printStackTrace(TAG, e);
            return null;
        }
    }
    
    /**
     * 结算能量最高的那一局
     */
    private static void settleBestRound(GameSession session, String source) {
        try {
            Log.runtime(TAG, "正在结算第" + session.roundNumber + "局游戏（能量最高）");
            
            JSONObject response = new JSONObject(AntForestRpcCall.settlementWhackMole(
                session.token, session.remainingMoleIds, source));
                
            if (ResChecker.checkRes(TAG, response)) {
                int totalEnergy = response.optInt("totalEnergy", 0);
                int provideEnergy = response.optInt("provideDefaultEnergy", 0);
                int whackedEnergy = totalEnergy - provideEnergy;
                
                Log.forest("森林能量⚡️[6秒拼手速第" + session.roundNumber + "局结算" + 
                          " 打地鼠能量+" + whackedEnergy + "g" +
                          " 默认奖励+" + provideEnergy + "g" +
                          " 总能量+" + totalEnergy + "g]");
            }
        } catch (Exception e) {
            Log.runtime(TAG, "结算第" + session.roundNumber + "局失败: " + e.getMessage());
            Log.printStackTrace(TAG, e);
        }
    }
    
    /**
     * 安全关闭线程池
     */
    private static void shutdownExecutor(ExecutorService executor) {
        executor.shutdown();
        try {
            if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
                executor.shutdownNow();
            }
        } catch (InterruptedException e) {
            executor.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
}

File: app/src/main/java/fansirsqi/xposed/sesame/task/antForest/WhackMole.kt
package fansirsqi.xposed.sesame.task.antForest

import android.annotation.SuppressLint
import fansirsqi.xposed.sesame.data.Status
import fansirsqi.xposed.sesame.hook.Toast
import fansirsqi.xposed.sesame.util.Log
import fansirsqi.xposed.sesame.util.ResChecker
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.max

/**
 * 6秒拼手速打地鼠
 * 整合版本：适配最新 RPC 定义
 */
object WhackMole {
    private const val TAG = "WhackMole"
    private const val SOURCE = "senlinguangchangdadishu"
    private const val EXEC_FLAG = "forest::whackMole::executed"

    @Volatile
    private var totalGames = 5
    private const val GAME_DURATION_MS = 12000L
    private val globalScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val startTime = AtomicLong(0)

    @Volatile
    private var isRunning = false

    enum class Mode {
        COMPATIBLE, // 兼容模式 (对应 old系列 RPC)
        AGGRESSIVE  // 激进模式 (对应 标准系列 RPC)
    }

    data class GameSession(
        val token: String,
        val roundNumber: Int
    )

    fun setTotalGames(games: Int) {
        totalGames = games
    }

    private val intervalCalculator = GameIntervalCalculator

    fun start(mode: Mode) {
        if (isRunning) {
            Log.record(TAG, "⏭️ 打地鼠游戏正在运行中，跳过重复启动")
            return
        }
        isRunning = true

        globalScope.launch {
            try {
                when (mode) {
                    Mode.COMPATIBLE -> runCompatibleMode()
                    Mode.AGGRESSIVE -> runAggressiveMode()
                }
                Status.setFlagToday(EXEC_FLAG)
            } catch (e: Exception) {
                Log.printStackTrace(TAG, "打地鼠异常: ", e)
            } finally {
                isRunning = false
                Log.record(TAG, "🎮 打地鼠运行状态已重置")
            }
        }
    }

    // ================= [ 兼容模式：对应 old 系列 RPC ] =================

    private suspend fun runCompatibleMode() = withContext(Dispatchers.IO) {
        try {
            val startTs = System.currentTimeMillis()

            // 1. 开始游戏 (使用 oldstartWhackMole)
            val response = JSONObject(AntForestRpcCall.oldstartWhackMole(SOURCE))
            if (!response.optBoolean("success")) {
                Log.record(TAG, response.optString("resultDesc", "开始失败"))
                return@withContext
            }

            val moleInfoArray = response.optJSONArray("moleInfo")
            val token = response.optString("token")
            if (moleInfoArray == null || token.isEmpty()) return@withContext

            val allMoleIds = mutableListOf<Long>()
            val bubbleMoleIds = mutableListOf<Long>()

            for (i in 0 until moleInfoArray.length()) {
                val mole = moleInfoArray.getJSONObject(i)
                val moleId = mole.getLong("id")
                allMoleIds.add(moleId)
                if (mole.has("bubbleId")) bubbleMoleIds.add(moleId)
            }

            // 2. 打有能量球的地鼠 (使用 oldwhackMole)
            var hitCount = 0
            bubbleMoleIds.forEach { moleId ->
                try {
                    val whackResp = JSONObject(AntForestRpcCall.oldwhackMole(moleId, token, SOURCE))
                    if (whackResp.optBoolean("success")) {
                        val energy = whackResp.optInt("energyAmount", 0)
                        hitCount++
                        Log.forest("森林能量⚡️[兼容打地鼠:$moleId +${energy}g]")
                        if (hitCount < bubbleMoleIds.size) {
                            delay(100 + (0..200).random().toLong())
                        }
                    }
                } catch (t: Throwable) {
                }
            }

            // 3. 计算剩余 ID 并结算 (使用 oldsettlementWhackMole)
            val remainingIds = allMoleIds.filter { !bubbleMoleIds.contains(it) }.map { it.toString() }
            val elapsedTime = System.currentTimeMillis() - startTs
            delay(max(0L, 6000L - elapsedTime - 200L))

            val settleResp = JSONObject(AntForestRpcCall.oldsettlementWhackMole(token, remainingIds, SOURCE))
            if (ResChecker.checkRes(TAG, settleResp)) {
                val total = settleResp.optInt("totalEnergy", 0)
                Log.forest("森林能量⚡️[兼容模式完成 总能量+${total}g]")
            }
        } catch (t: Throwable) {
            Log.record(TAG, "兼容模式出错: ${t.message}")
        }
    }

    // ================= [ 激进模式：对应 标准系列 RPC ] =================

    @SuppressLint("DefaultLocale")
    private suspend fun runAggressiveMode() {
        startTime.set(System.currentTimeMillis())
        val dynamicInterval = intervalCalculator.calculateDynamicInterval(GAME_DURATION_MS, totalGames)

        val sessions = mutableListOf<GameSession>()
        try {
            for (roundNum in 1..totalGames) {
                // 1. 启动单局 (使用标准 startWhackMole)
                val session = startSingleRound(roundNum)
                if (session != null) sessions.add(session)

                if (roundNum < totalGames) {
                    val remaining = GAME_DURATION_MS - (System.currentTimeMillis() - startTime.get())
                    delay(intervalCalculator.calculateNextDelay(dynamicInterval, roundNum, totalGames, remaining))
                }
            }
        } catch (e: CancellationException) {
            return
        }

        // 等待结算窗口
        val waitTime = max(0L, GAME_DURATION_MS - (System.currentTimeMillis() - startTime.get()))
        delay(waitTime)

        // 2. 批量结算 (使用标准 settlementWhackMole)
        var totalEnergy = 0
        sessions.forEach { session ->
            delay(200)
            totalEnergy += settleStandardRound(session)
        }
        Log.forest("森林能量⚡️[激进模式${sessions.size}局 总计${totalEnergy}g]")
    }

    private suspend fun startSingleRound(round: Int): GameSession? = withContext(Dispatchers.IO) {
        try {
            // 标准接口调用
            val startResp = JSONObject(AntForestRpcCall.startWhackMole())
            if (!ResChecker.checkRes(TAG, startResp)) return@withContext null

            if (!startResp.optBoolean("canPlayToday", true)) {
                Status.setFlagToday(EXEC_FLAG)
                throw CancellationException("Today limit reached")
            }

            val token = startResp.optString("token")
            Toast.show("打地鼠 第${round}局启动\nToken: $token")
            GameSession(token, round)
        } catch (e: Exception) {
            null
        }
    }

    private suspend fun settleStandardRound(session: GameSession): Int = withContext(Dispatchers.IO) {
        try {
            // 标准结算调用 (RPC 内部会自动处理 moleIdList 1-15)
            val resp = JSONObject(AntForestRpcCall.settlementWhackMole(session.token))
            if (ResChecker.checkRes(TAG, resp)) {
                return@withContext resp.optInt("totalEnergy", 0)
            }
        } catch (e: Exception) {
        }
        0
    }
}